Full name: Nguyễn Bửu Thạch

Student ID: 20120576

# HW2: Parallel Execution in CUDA

**To compile your file, you can use this command:** \
`nvcc tên-file.cu -o tên-file-chạy` \
***You can use Vietnamese to anwser the questions***

## Question 1A

In [ ]:
!nvcc HW2_P1.cu -o HW2_P1

In [ ]:
!./HW2_P1

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 bytes
****************************

Input size: 16777217
Host time: 61.411583 ms

Kernel 1
Grid size: 8193, block size: 1024
Kernel time = 2.904512 ms
CORRECT :)

Kernel 2
Grid size: 8193, block size: 1024
Kernel time = 3.141568 ms
CORRECT :)

Kernel 3
Grid size: 8193, block size: 1024
Kernel time = 0.164064 ms
INCORRECT :(


Từ kết quả chạy, bạn so sánh “kernel time” của 3 hàm kernel với nhau và thử giải thích
xem tại sao lại như vậy.\
Không biết tại sao

In [ ]:
!nvprof ./HW2_P1

==594== NVPROF is profiling process 594, command: ./HW2_P1
**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 bytes
****************************

Input size: 16777217
Host time: 43.271713 ms

Kernel 1
Grid size: 8193, block size: 1024
Kernel time = 2.902944 ms
CORRECT :)

Kernel 2
Grid size: 8193, block size: 1024
Kernel time = 2.754752 ms
CORRECT :)

Kernel 3
Grid size: 8193, block size: 1024
Kernel time = 0.231456 ms
INCORRECT :(
==594== Profiling application: ./HW2_P1
==594== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   75.79%  149.58ms         3  49.859ms  47.623ms  51.895ms  [CUDA memcpy DtoH]
                   21.44%  42.318ms         3  14.106ms  13.920ms  14.417ms  [CUDA memcpy HtoD]
                    1.36%  2.6892ms         1  2.6892ms  2.6892ms  2.6892ms  reduceBlksKernel2(int*, int*, int)
           

Nhận xét về thời gian thực thi các công việc trong phần “GPU activities” \
- Thời gian truyền dữ liệu giữa GPU và CPU chiếm phần lớn thời gian, điều này cho thấy việc truyền dữ liệu giữa CPU và GPU là yếu tố quan trọng trong quá trình.
- Các kenel thì thực thi nhanh chóng , chỉ chiếm 1 phần nhỏ thời gian.

## Question 1B


In [ ]:
!./HW2_P1 1024

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 bytes
****************************

Input size: 16777217
Host time: 46.011295 ms

Kernel 1
Grid size: 8193, block size: 1024
Kernel time = 2.819072 ms
CORRECT :)

Kernel 2
Grid size: 8193, block size: 1024
Kernel time = 2.718688 ms
CORRECT :)

Kernel 3
Grid size: 8193, block size: 1024
Kernel time = 0.154816 ms
INCORRECT :(


In [ ]:
!./HW2_P1 512

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 bytes
****************************

Input size: 16777217
Host time: 43.433216 ms

Kernel 1
Grid size: 16385, block size: 512
Kernel time = 2.117504 ms
CORRECT :)

Kernel 2
Grid size: 16385, block size: 512
Kernel time = 2.027072 ms
CORRECT :)

Kernel 3
Grid size: 16385, block size: 512
Kernel time = 0.157952 ms
INCORRECT :(


In [ ]:
!./HW2_P1 256

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 bytes
****************************

Input size: 16777217
Host time: 50.593151 ms

Kernel 1
Grid size: 32769, block size: 256
Kernel time = 1.839104 ms
CORRECT :)

Kernel 2
Grid size: 32769, block size: 256
Kernel time = 1.644864 ms
CORRECT :)

Kernel 3
Grid size: 32769, block size: 256
Kernel time = 0.159392 ms
INCORRECT :(


In [ ]:
!./HW2_P1 128

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 bytes
****************************

Input size: 16777217
Host time: 43.191296 ms

Kernel 1
Grid size: 65537, block size: 128
Kernel time = 1.575488 ms
CORRECT :)

Kernel 2
Grid size: 65537, block size: 128
Kernel time = 1.402944 ms
CORRECT :)

Kernel 3
Grid size: 65537, block size: 128
Kernel time = 0.203744 ms
INCORRECT :(


Block size | Grid size | Num blocks / SM | Occupancy (%) | Kernel time (ms)
--- | --- | --- | --- | ---
1024 |  8193 | 1 | 100 | 2.819072
512 | 16385 | 2 | 100 | 2.117504
256 | 32769 | 4 | 100 | 1.839104
128 | 65537 | 8 | 100 | 1.575488

Theo ràng buộc từ compute capability 7.5
- Số block tối đa mỗi SM: 16
- Số thread tối đa mỗi SM: 1024 \

####Block size = 1024
- Số thread mỗi block: 1024
- Giới hạn thread trên mỗi SM: 1024

Suy ra:
- Num Blocks / SM = $\frac{1024}{1024}$ = 1
- Occupancy (%) = 100% (do tổng số thread đạt giới hạn 1024 thread cho mỗi SM).

####Block size = 512
- Số thread mỗi block: 512
- Giới hạn thread trên mỗi SM: 1024

Suy ra:
- Num Blocks / SM = $\frac{1024}{512}$ = 2
- Occupancy (%) = 100% (do tổng số thread đạt giới hạn 1024 thread cho mỗi SM).

####Block size = 256
- Số thread mỗi block: 1024
- Giới hạn thread trên mỗi SM: 1024

Suy ra:
- Num Blocks / SM = $\frac{1024}{256}$ = 4
- Occupancy (%) = 100% (do tổng số thread đạt giới hạn 1024 thread cho mỗi SM).

####Block size = 128
- Số thread mỗi block: 128
- Giới hạn thread trên mỗi SM: 1024

Suy ra:
- Num Blocks / SM = $\frac{1024}{128}$ = 8
- Occupancy (%) = 100% (do tổng số thread đạt giới hạn 1024 thread cho mỗi SM).


## Question 2A

In [ ]:
!nvcc HW2_P2.cu -o HW2_P2

In [ ]:
!./HW2_P2

**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 bytes
****************************

Processing time (use host): 3541.761963 ms

Basic Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 8.573024 ms
Error between device result and host result: 0.000004

Shared memory Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 5.461952 ms
Error between device result and host result: 0.000004



In [ ]:
!nvprof ./HW2_P2

==2891== NVPROF is profiling process 2891, command: ./HW2_P2
**********GPU info**********
Name: Tesla T4
Compute capability: 7.5
Num SMs: 40
Max num threads per SM: 1024
Max num warps per SM: 32
GMEM: 15835660288 bytes
****************************

Processing time (use host): 3700.514893 ms

Basic Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 8.440320 ms
Error between device result and host result: 0.000004

Shared memory Matrix Multiplication:
Grid size: 32 * 32, block size: 32 * 32
Processing time (use device): 5.295264 ms
Error between device result and host result: 0.000004

==2891== Profiling application: ./HW2_P2
==2891== Profiling result:
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name
 GPU activities:   34.72%  3.2667ms         1  3.2667ms  3.2667ms  3.2667ms  matrix_multiplication_kernel1(float*, float*, float*, int, int, int)
                   28.22%  2.6551ms         1  2.6551ms  2.6551ms  2.65

Nhận xét về thời gian thực thi các công việc trong phần “GPU activities” \
- Các kenel thực thi nhanh chóng và chiếm phần lớn thời gian, điều này cho thấy việc tính toán nhân ma trận đã được thực hiện một cách hiệu quả.
- Thời gian truyền dữ liệu giữa GPU và CPU chiếm phần nhỏ hơn so với thời gian thực thi các kernel, không còn chiếm nhiều thời gian trong quá trình nữa

## Question 2B

**For Basic Matrix Multipication**

$m= 2^{10} = 1024$ \
$n= 2^9 = 512$ \
$k= 2^{10} = 1024$


1. How many floating operations are being performed in your matrix multiply
kernel? Explain

Với ma trận A kích thước $𝑚*n$ và ma trận B kích thước $n*k$: \
- Mỗi phần tử trong ma trận kết quả C (kích thước $𝑚*k$) cần $n$ phép nhân và $n-1$ phép cộng
- Số phép toán cho toàn bộ ma trận C: \
Phép nhân: $m*k*n$ \
Phép nhân: $m*k*(n-1)$ \

Vậy tổng số FLOPs cần cho toàn bộ phép nhân ma trận là $m*k*n + m*k*(n-1) = m*k*(2n-1) = 1024*1024*(2*512-1)$ \


2. How many global memory reads are being performed by your kernel? Explain

Với ma trận A kích thước $m*n$ và ma trận B kích thước $n*k$:
- Mỗi phần tử của C ta cần n phép đọc từ A và n phép đọc từ B \
- Ma trận C(kích thước $m*k$) có tổng cộng $m*k$ phần tử

Do đó: \
- Tổng số phép đọc từ A là $m*k*n$ \
- Tổng số phép đọc từ B là $m*k*n$ \

Vậy tổng số lần đọc bộ nhớ toàn cục (global memory) là \
$m*k*n+m*k*n = 2*m*n*k = 2*1024*512*1024$

3. How many global memory writes are being performed by your kernel? Explain.

- Mỗi phần tử của ma trận C chỉ được ghi vào bộ nhớ toàn cục một lần sau khi hoàn thành tính toán.
- Ma trận C kích thước $m*k$, nghĩa là ma trận C có $m*k$ phần tử  \

Tổng số phép ghi vào bộ nhớ toàn cục là
$m*k=1024*1024$


**For Tiled Matrix Multipication**

1. How many floating operations are being performed in your matrix multiply
kernel? Explain

2. How many global memory reads are being performed by your kernel? Explain

3. How many global memory writes are being performed by your kernel? Explain